drivers/SmartThings/matter-switch/src/init.lua

```diff
@@ -49,23 +51,65 @@ local function find_default_endpoint(device, component)
   return res
 end
 
-local function component_to_endpoint(device, component_id)
-  -- Assumes matter endpoint layout is sequentional starting at 1.
-  local ep_num = component_id:match("switch(%d)")
-  return ep_num and tonumber(ep_num) or find_default_endpoint(device, component_id)
+local function initialize_switch(device)
+  local switch_eps = device:get_endpoints(clusters.OnOff.ID)
+  table.sort(switch_eps)
+
+  local endpoint_map = {}
+  local current_component_number = 1
+
+  -- For switch devices, the profile components follow the naming convention "switch%d",
+  -- with the exception of "main" being the first component. Each component will then map
+  -- to the next lowest endpoint that hasn't been mapped yet.
+  for _, ep in ipairs(switch_eps) do
+    if current_component_number == 1 then
+      endpoint_map[ep] = "main"
+    else
+      endpoint_map[ep] = string.format("switch%d", current_component_number)
+    end
+    current_component_number = current_component_number + 1
+  end
+
+  device:set_field(ENDPOINT_TO_COMPONENT_MAP, endpoint_map, {persist = true})
+
+  -- New profiles need to be added for devices that have more switch endpoints
+  local MAX_MULTI_SWITCH_EPS = 7
+  -- Note: This profile switching is needed because of shortcoming in the generic fingerprints
+  -- where devices with multiple endpoints with the same device type cannot be detected
+  local num_switch_eps = #switch_eps
+  if num_switch_eps > 1 then
+    device:try_update_metadata({profile = string.format("switch-%d", math.min(num_switch_eps, MAX_MULTI_SWITCH_EPS))})
+  end
+  if num_switch_eps > MAX_MULTI_SWITCH_EPS then
+    error(string.format(
+      "Matter multi switch device will have limited function. Profile doesn't exist with %d components",
+      num_switch_eps
+    ))
+  end
+end
+
+local function component_to_endpoint(device, component_name)
+  local map = device:get_field(ENDPOINT_TO_COMPONENT_MAP) or {}
+  for ep, component in pairs(map) do
+    if component == component_name then return ep end
+  end
+  return find_default_endpoint(device, component_name)
 end
 
 local function endpoint_to_component(device, ep)
-  local switch_comp = string.format("switch%d", ep)
-  if device.profile.components[switch_comp] ~= nil then
-    return switch_comp
-  else
-    return "main"
+  local map = device:get_field(ENDPOINT_TO_COMPONENT_MAP) or {}
+  if map[ep] and device.profile.components[map[ep]] then
+    return map[ep]
   end
+  return "main"
```

Comment:
I think what we probably want to do here is, if we have _any_ endpoint to component mapping done at all, to ignore messages from a non-mapped endpoint, rather than emit them from main

I'm thinking of a situation where you have more switches than we have profiles for. Anything from endpoint > X will change the state of "main" which is also a _real_ component that we have mapped to.

Reply:
The `emit_event_for_endpoint` function expects the `endpoint_to_component` function to return something, or else we would run into an issue in `emit_component_event` when we try to use the nil value for the component. So, I think there needs to be some kind of default value, even if it is not ideal.

> I'm thinking of a situation where you have more switches than we have profiles for. Anything from endpoint > X will change the state of "main" which is also a real component that we have mapped to.

I think this is a valid problem that would be caused by this. The solution would be to add another profile to account for the number of switches. Ideally we'd be able to quickly see the error in the logs since it is logged if a device has more than our supported number of switches

Reply:
if you feel like digging into it, what I did with buttons was use parent-child in the case that I didn't have a profile configured for that number of components